### PR TITLE
Unzip CWL Dependencies and make available to pre-processing

### PR DESCRIPTION
**Pull request by danbills**
I'm throwing in some nice-to-haves here:

* deleted unused "mungeId" artifact before FullyQualifiedName was born
* added "sbt-pack" SBT plugin which makes nice run scripts and doesn't assemble fat jar (slow)

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala [hunk 1]
```diff
@@ -434,22 +435,26 @@ class MaterializeWorkflowDescriptorActor(serviceRegistryActor: ActorRef,
...
-      for {
-        cwl <- CwlDecoder.decodeAllCwl(cwlFile)
-        wf <- fromEither[IO](cwl.select[Workflow].toRight(NonEmptyList.one(s"expected a workflow but got a $cwl")))
-        executable <-  fromEither[IO](wf.womExecutable(Option(source.inputsJson)))
-        graph <- fromEither[IO](executable.graph.toEither)
-        ioFunctions = new WdlFunctions(pathBuilders)
-        validatedWomNamespace <- fromEither[IO](validateWomNamespace(executable, ioFunctions))
-      } yield validatedWomNamespace
-    } finally {
-      cwlFile.delete(swallowIOExceptions = true)
```
Comment (danbills) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: This `finally` block was for some reason executing before the parsing step, throwing `FileNotFound` error from `cwltool`.  

The file and the dependencies are being created in a temporary directory though, and they're small, so I don't think this is that bad.

I am thinking of another way to ensure they get deleted but short of using fs2 I don't know how to ensure resource safety without the (problematic) `finally` block.

**Comment by danbills** [comment]
For the record I've tested this manually.

File: cwl/src/main/scala/cwl/GlobEvaluator.scala [hunk 2]
```diff
@@ -51,7 +51,8 @@ object GlobEvaluator {
             case WomString(value) => Vector(value)
             case WomArray(WomArrayType(WomStringType), values) => values.map(_.valueString)
             case result =>
-              throw new RuntimeException(s"TODO: WOM: Placeholder exception: unexpected expression result: $result")
+              throw new RuntimeException(
+                s"TODO: WOM: Placeholder exception: unexpected expression result: $result on input $ecmaScript using inputs $parameterContext")
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/GlobEvaluator.scala]: I'll make it more beauteous.  Was debugging a... bug.

File: cwl/src/test/resources/tar-param.cwl [hunk 3]
```diff
@@ -1,21 +1,17 @@
...
-baseCommand: 
-  - "tar"
-  - "xf"
+baseCommand: ["tar", "xf"]
 inputs: 
   - type: "string"
     inputBinding: 
       position: 2
-    id: "file:///home/dan/wdl4s/tar-param.cwl#extractfile"
+    id: "extractfile"
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/test/resources/tar-param.cwl]: to make sure the pre-processor is pre-processing (and thus supplying) all this stuff correctly.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala [hunk 4]
```diff
@@ -434,22 +435,26 @@ class MaterializeWorkflowDescriptorActor(serviceRegistryActor: ActorRef,
...
+
+    val cwlFile: File = File.newTemporaryFile(prefix = workflowIdForLogging.toString).write(source.workflowSource)
+
+    def unzipDependencies: Parse[Unit]  = source match {
+      case wsfwdz: WorkflowSourceFilesWithDependenciesZip =>
+        EitherT.
+          fromEither[IO]{validateImportsDirectory(wsfwdz.importsZip, Some(DefaultPathBuilder.build(cwlFile.parent.path))).toEither}.
+          map(_ => ())
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: I tried that :(... maybe I'm missing an import.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala [hunk 5]
```diff
@@ -434,22 +435,26 @@ class MaterializeWorkflowDescriptorActor(serviceRegistryActor: ActorRef,
...
+      case _ => Monad[Parse].unit
     }
+
+    for {
+      _ <- unzipDependencies
+      cwl <- CwlDecoder.decodeAllCwl(cwlFile)
+      wf <- fromEither[IO](cwl.select[Workflow].toRight(NonEmptyList.one(s"expected a workflow but got a $cwl")))
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: In my defense this is all pre-existing, just deleted the `try/finally` block and added the first line.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala [hunk 6]
```diff
@@ -434,22 +435,26 @@ class MaterializeWorkflowDescriptorActor(serviceRegistryActor: ActorRef,
...
+      executable <-  fromEither[IO](wf.womExecutable(Option(source.inputsJson)))
+      graph <- fromEither[IO](executable.graph.toEither)
+      ioFunctions = new WdlFunctions(pathBuilders)
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: Pre-existing.  This should probably be addressed in another PR.

File: project/plugins.sbt [hunk 7]
```diff
@@ -8,3 +8,4 @@ See https://github.com/broadinstitute/cromwell/issues/645
 addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "0.7.1")
 addSbtPlugin("com.github.gseitz" % "sbt-release" % "1.0.3")
 addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.5.0")
+addSbtPlugin("org.xerial.sbt" % "sbt-pack" % "0.9.1")
```
[the remark this answers is not in the document]
Comment (danbills) [reply project/plugins.sbt]: It's super nice for my current manual testing arrangement.

* creates a script to run `java -jar X.jar` for all runnable mainclasses
* not creating fat jar saves a couple dozen seconds

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: `import cats.syntax.functor._` I think

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: I mean I love the burn 😄

File: project/plugins.sbt (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply project/plugins.sbt]: Those sound like very nice things indeed. 🙂

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (Horneth) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: it's just a naming issue, `WdlFunctions` should be renamed `IoFunctions`

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 1 above)
Comment (Horneth) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: yeah I'm surprised the finally runs before the try Oo
We do delete the files for WDL for workflows though so it'd be nice to have at least a trace as to why it's not the case for CWL ones if we leave it like that

**Comment by Horneth** [comment]
👍

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 1 above)
Comment (danbills) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: I just figured it out.  It's creating an `IO` ultimately, so not running it!  Thus the finally block deletes the file before the code is actually executed.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: It should also be moved out of the `wdl` subproject, but it looks like that should be straightforward.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala (hunk 1 above)
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala]: Nice catch! I'm so used to the Future paradigm that runs immediately. 😄

**Pull request merged**